design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,290 @@
+
+Configuration parameters:
+1) Heartbeat INTERVAL (e.g send heartbeat every 200 ms)
+2) Heartbeat WINDOW (e.g every 2 seconds, count the missed/received heartbeats and determine if server is down/up)
+3) MISSED_THRESHOLD: How many heartbeats in a row constitute a node as down (e.g. 3 missed heartbeats = server is down)
+4) RECEIVED_THRESHOLD: How many heartbeats in a row constitute a node as up (e.g. 2 received heartbeats = server is up)
+
+We will provide sane defaults out-of-box, to achieve a 99% uptime.
+
+Pseudocode for REST endpoint for `/heartbeat`:
+```java
+Map<KsqlServer, List<Long>> receivedHeartbeats;
+
+@POST
+public Response receiveHeartbeat(Request heartbeat) 
+  return Response.ok(processRequest()).build();
+}
+
+private HeartbeatResponse processRequest() {
+  KsqlServer server = request.getServer();
+  receivedHeartbeats.get(server).add(System.currentTimeMillis());
+}
+```
+
+
+### Lag reporting
+
+Every server neeeds to periodically broadcast their local lag information. We will implement a new 
+REST endpoint `/reportlag`. The lag information at a server is obtained via 
+`Map<String, Map<Integer, Long>> localLagMap = kafkaStreams.allLocalOffsetLags();`.
+
+Pseudocode for REST endpoint for `/reportlag`:
+```java
+Map<KsqlServer, Map<String, Map<Integer, Long>> globalLagMap;
+
+@POST
+public Response receiveLagInfo(Request lagMap) 
+  return Response.ok(processRequest()).build();
+}
+
+private LagReportResponse processRequest() {
+  KsqlServer server = request.getServer();
+  Map<String, Map<Integer, Long>> localLagMap = request.getLag();
+  globalLagMap.put(server, localLagMap);
+}
+``` 
+
+### Lag-aware routing
+
+Given the above information (alive + lag), how does a KSQL server decide to which standby to route 
+the request? Every server knows the lag of all partitions hosted by all servers in the cluster. 
+This allows us to implement lag-aware routing where server `B` can a) determine that server `A` is 
+down and b) decide whether it will serve the request itself or forward it to `C` depending on who 
+has the smallest lag for the given key in the pull query. 
+
+Pseudocode for routing:
+```java
+// Map populated periodically through heartbeat information
+Map<KsqlNode, Boolean> aliveNodes;
+
+// Map populated periodically through lag reporting
+Map<KsqlNode, Map<String, Map<Integer, Long>> globalLagMap;
+ 
+// Fetch the metadata related to the key
+KeyQueryMetadata queryMetadata = queryMetadataForKey(store, key, serializer);
+ 
+// Acceptable lag for the query
+final long acceptableOffsetLag = 10000;
+
+// Ordered list of servers, in order of most caught-up to least
+List<KsqlNode> nodesToQuery;
+KsqlNode active = queryMetadata.getActiveHost();
+if (aliveNodes.get(active)) {
+  nodesToQuery.add(active)  
+}
+
+// add standbys
+nodesToQuery.addAll(queryMetadata.getStandbyHosts().stream()
+    // filter out all the standbys that are down
+    .filter(standbyHost -> aliveNodes.get(standByHost) == null)
```

Comment:
I believe the the java stream filter retains anything that passes the predicate.  It doesn't remove things that don't pass: https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#filter-java.util.function.Predicate-

Reply:
> the java stream filter retains anything that passes the predicate. It doesn't remove things that don't pass

Isn't the latter just another way of saying the former? :D :p

If think you mean "It doesn't remove things that pass'.  Code should be:

```java
.filter(standbyHost -> aliveNodes.get(standByHost) != null)
```